packages/core/contracts/financial-templates/common/FundingRatePayer.sol

```diff
@@ -0,0 +1,121 @@
+     * in a week or more then a late penalty is applied which is sent to the caller. If the amount of
+     * fees owed are greater than the pfc, then this will pay as much as possible from the available collateral.
+     * An event is only fired if the fees charged are greater than 0.
+     * @return totalPaid Amount of collateral that the contract paid (sum of the amount paid to the Store and caller).
+     * This returns 0 and exit early if there is no pfc, fees were already paid during the current block, or the fee rate is 0.
+     */
+    function payFundingRateFees() public nonReentrant() returns (FixedPoint.Unsigned memory totalPaid) {
+        FundingRateStoreInterface fundingRateStore = FundingRateStoreInterface(
+            finder.getImplementationAddress("FundingRateStore")
+        );
+        uint256 time = getCurrentTime();
+        FixedPoint.Unsigned memory collateralPool = _pfc();
+
+        // Exit early if there is no collateral from which to pay fees.
+        if (collateralPool.isEqual(0)) {
+            return totalPaid;
+        }
+
+        // Exit early if fees were already paid during this block.
+        if (lastPaymentTime == time) {
+            return totalPaid;
+        }
+
+        (FixedPoint.Unsigned memory fundingRateFee, FixedPoint.Unsigned memory latePenalty) = fundingRateStore
+            .computeFundingRateFee(lastPaymentTime, time, collateralPool);
+        lastPaymentTime = time;
+
+        totalPaid = fundingRateFee.add(latePenalty);
+        if (totalPaid.isEqual(0)) {
+            return totalPaid;
+        }
+        // If the effective fees paid as a % of the pfc is > 100%, then we need to reduce it and make the contract pay
+        // as much of the fee that it can (up to 100% of its pfc). We'll reduce the late penalty first and then the
+        // regular fee, which has the effect of paying the store first, followed by the caller if there is any fee remaining.
+        if (totalPaid.isGreaterThan(collateralPool)) {
+            FixedPoint.Unsigned memory deficit = totalPaid.sub(collateralPool);
+            FixedPoint.Unsigned memory latePenaltyReduction = FixedPoint.min(latePenalty, deficit);
+            latePenalty = latePenalty.sub(latePenaltyReduction);
+            deficit = deficit.sub(latePenaltyReduction);
+            fundingRateFee = fundingRateFee.sub(FixedPoint.min(fundingRateFee, deficit));
+            totalPaid = collateralPool;
+        }
+
+        emit FundingRateFeesPaid(fundingRateFee.rawValue, latePenalty.rawValue);
+
+        _adjustCumulativeFeeMultiplier(totalPaid, collateralPool);
+
+        if (fundingRateFee.isGreaterThan(0)) {
+            collateralCurrency.safeIncreaseAllowance(address(fundingRateStore), fundingRateFee.rawValue);
+            fundingRateStore.payFundingRateFees(fundingRateFee);
+        }
+
+        if (latePenalty.isGreaterThan(0)) {
+            collateralCurrency.safeTransfer(msg.sender, latePenalty.rawValue);
```

Comment:
who is the sender here? if it is an EOA why would they ever call this function

Reply:
This is paid in the fees() modifier. But yeah you're right no EOA should be incentivized to call it. It mirrors the FeePayer interface

Reply:
Wait, what? The contract is _paying_ the EOA, right? So the late fee _is_ what incentivizes the EOA to pay IIUC.

Reply:
Ah you are right @mrice32 , everyone is incentivized to call this function if `latePenalty` is ever > gas fees

Reply:
i see i misunderstood what was going where with safeTransfer.